src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpUploadHandler.java

```diff
@@ -0,0 +1,109 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.remote.blobstore.http;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.auth.Credentials;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.DefaultHttpRequest;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpChunkedInput;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.stream.ChunkedStream;
+import io.netty.util.internal.StringUtil;
+
+class HttpUploadHandler extends AbstractHttpHandler<FullHttpResponse> {
+
+  public HttpUploadHandler(Credentials credentials) {
+    super(credentials);
+  }
+
+  @Override
+  protected void channelRead0(ChannelHandlerContext ctx, FullHttpResponse response)
+      throws Exception {
+    try {
+      checkState(userPromise != null, "response before request");
+      if (!response.status().equals(HttpResponseStatus.OK)
+          && !response.status().equals(HttpResponseStatus.ACCEPTED)
+          && !response.status().equals(HttpResponseStatus.CREATED)
+          && !response.status().equals(HttpResponseStatus.NO_CONTENT)) {
+        // Supporting more than OK status to be compatible with nginx webdav.
+        failAndResetUserPromise(
+            new HttpException(
+                response.status(), "Download failed with " + "Status: " + response.status(), null));
+      } else {
+        succeedAndResetUserPromise();
+      }
+    } finally {
+      if (!HttpUtil.isKeepAlive(response)) {
+        ctx.close();
+      }
+    }
+  }
+
+  @Override
+  public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+      throws Exception {
+    checkState(userPromise == null, "handler can't be shared between pipelines.");
```

Comment:
This isn't thread-safe. Does it need to be? Maybe make userPromise an AtomicReference and use a compre-and-swap here?

Reply:
No it doesn't need to be thread-safe. There is no parallelism within a channel handler.